include/flashinfer/comm/trtllm_mnnvl_allreduce.cuh

```diff
@@ -95,581 +96,1126 @@ inline __device__ __nv_half fromFloat<__nv_half>(float val) {
+  T* scatterBufDest =
+      reinterpret_cast<T*>(flag.getCurLamportBuf(inputPtrs[destRank], MNNVLTwoShotStage::SCATTER));
+  T* broadcastBufW =
+      reinterpret_cast<T*>(flag.getCurLamportBuf(mcastPtr, MNNVLTwoShotStage::BROADCAST));
+  T* broadcastBufR =
+      reinterpret_cast<T*>(flag.getCurLamportBuf(inputPtrs[rank], MNNVLTwoShotStage::BROADCAST));
 
-        if (LAMPORT) {
-          // Assume that the 16B were written atomically, so we only need to check one value
-          T_IN lowest_val = *(T_IN*)&value;
-          valid &= !isNegZero(lowest_val);
-        }
-        *dst4 = value;
-      }
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
+  cudaTriggerProgrammaticLaunchCompletion();
+#endif
+  // Make sure the clear function is called before OOB thread exits
+  if (packedIdx * kELTS_PER_THREAD >= tokenDim) {
+    flag.clearDirtyLamportBuf(inputPtrs[rank], -1);
+    return;
+  }
+
+  // =============================== Scatter ===============================
+
+  // Load vectorized data
+  PackedVec<PackedType, T> val;
+  val.packed = loadPacked<PackedType>(&shardPtr[threadOffset]);
+#pragma unroll
+  for (int i = 0; i < kELTS_PER_THREAD; i++) {
+    if (isNegZero(val.elements[i])) {
+      val.elements[i] = fromFloat<T>(0.F);
     }
   }
 
-  __syncthreads();
+  // Store vectorized data
+  reinterpret_cast<PackedType*>(
+      &scatterBufDest[destTokenOffset * tokenDim * WorldSize + rank * tokenDim])[packedIdx] =
+      val.packed;
 
-  // Perform the initial input reduction
-  if (NUM_INPUTS > 0) {
-    T_IN accum[ELTS_PER_THREAD];
-    float4* accum4 = (float4*)&accum;
+  flag.clearDirtyLamportBuf(inputPtrs[rank], MNNVLTwoShotStage::SCATTER);
 
-    for (int j = 0; j < DIM / (ELTS_PER_THREAD * NUM_THREADS); j++) {
-      int k = j * NUM_THREADS + threadIdx.x;
+  // =============================== Reduction and Broadcast ===============================
 
-      *accum4 = *(float4*)&sh_input[k * ELTS_PER_THREAD];
+  if ((token % WorldSize) == rank) {
+    int localToken = token / WorldSize;
+    float accum[kELTS_PER_THREAD] = {0.F};
+
+    // Use float as we only check each float value for validity
+    PackedVec<PackedType, float> valuesLamport[WorldSize];
+    while (1) {
+      bool valid = true;
+#pragma unroll
+      for (int r = 0; r < WorldSize; r++) {
+        valuesLamport[r].packed = loadPackedVolatile<PackedType>(
+            &scatterBufLocal[localToken * tokenDim * WorldSize + r * tokenDim +
+                             packedIdx * kELTS_PER_THREAD]);
 
-      for (int i = 1; i < NUM_INPUTS; i++) {
-        float4 data = *(float4*)&sh_input[i * NUM_THREADS * ITERS + k * ELTS_PER_THREAD];
-        T_IN* p_d = (T_IN*)&data;
-        for (int x = 0; x < ELTS_PER_THREAD; x++) {
-          accum[x] += p_d[x];
+        // Check validity across all elements
+#pragma unroll
+        for (int i = 0; i < kLAMPORT_ELTS_PER_PACKED; i++) {
+          valid &= !isNegZero(valuesLamport[r].elements[i]);
         }
```

Comment:
As long as one of the values in WorldSize isNegZero, we will keep reading the others over and over again. Perhaps these are all cache hits, in which case: ignore my comment. Otherwise, would it make sense to break out of the inner loop (kLAMPORT_ELTS_PER_PACKED) early?